.php-cs-fixer.dist.php

```diff
@@ -14,6 +14,7 @@
 
 use PhpCsFixer\Config;
 use PhpCsFixer\Finder;
+use PhpCsFixer\Runner\Parallel\ParallelConfig;
 
 $header = <<<'EOF'
```

Comment:
We can introduce the way for providing allowed memory limit and provide reasonable default. Personally I've added `ini_set('memory_limit', '-1');` to the `.php-cs-fixer.php` 😅. I am not sure if it's something that should be resolved on our side (at least not necessarily _now_).

Reply:
I have vanilla PHP installed without any custom configuration, and by running the fixer on fixer repo with default fix command I get memory error.

I understand you do not, because of your local change that is not committed.

I would like to have this experience fixed for whoever contributing to repo with default PHP config, rather than all-but-you getting need to overcome default memory limit.

Reply:
Please share your idea about it - how would you want to achieve this? I believe this is out of the scope of this PR, as you could hit memory limit with sequential analysis too. I don't want to put `ini_set()` anywhere, and also I don't want to make this PR even bigger with properly configurable memory limit. Tell me how you see it 🙂.

Reply:
imagine I'm John Doe without any clue what's going on.
I run Fixer with before this PR (aka sequential), everything works smooth.
I update the Fixer to version after merge of this PR, I run it again (now with parallel) and boom, I get memory crash. Tool is not working as expected for me and I put the blame on great feature, because it simply uses more memory.

if use of parallel requires more memory, maybe we can understand how much more memory and if that's really needed - and either remove the need for that extra memory or auto-increase memory limit (even in entry point of `php-cs-fixer` file)

Reply:
I will try to profile this, maybe there's something that can be cleared during the process.

Reply:
FYI I was able to profile the parallel runner:

<img width="1728" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/e036fe9e-0a6c-48df-865b-595659d1a654">

<img width="1728" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/933d2365-2d76-467a-aa3d-e10690237156">

I still need to analyse it, but it seems like memory usage is getting bigger and bigger through the analysis, so I suspect it's about React communication and data collected from workers. I'll try to find exact places where I could make optimisations.

Reply:
@staabm you're the Blackfire master, maybe you can give some tips where/how to look 😁.

Reply:
I can see something useful in the screenshots yet.

is the screenshot from the parent-process or from one of the parallel worker processes?
the first screenshot might indicate that memory is never freed. does the memory consumption stablize when you add `gc_collect_cycles` calls - e.g. after each file?

I would expect that once a file is processes all datastructures related to CS-fixing said file could be free'd.

Reply:
The profile is from the main process. I also think I need to free some memory, I'll check `gc_collect_cycles` 🙂.

> how to reproduce the profiles?

The easiest way is just checkout this branch, add `->setParallelConfig(ParallelConfig::detect())` to the config and run the analysis (in my case it was `blackfire run php ./php-cs-fixer check --verbose`).

Reply:
> I also think I need to free some memory, I'll check `gc_collect_cycles`

hopefully you don't. but `gc_collect_cycles` is good way for debugging. if it can cleanup memory it might tell us, that we have some memory cycles or simlar somewhere, which php could cleanup even without `gc_collect_cycles`, which would be best.

it might indicate something similar as https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types#lower-memory-consumption

Reply:
> The profile is from the main process.

I wonder why the main process is doing tokenizing? wasn't the point of this PR to do tokenizing in the workers?

Reply:
Here's a run with `gc_collect_cycles()` after each workers' callback. I can't remember when I used `goto` last time 😆.

<img width="1728" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/a5713668-3c57-446d-9df6-27c440ccfc70">

<img width="1098" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/815ec7c2-baa3-429f-94e6-e35d98d69a02">

Unfortunately, no improvement. Enough for now, I'll take a deeper look later.

Reply:
first thing I would try is making closures `static` whereever possible. it might be useful to pass `$this` via `use($self)` into closures to make sure php can free the memory of the closures and all referenced objects early enough

Reply:
For comparison, here's a profile from sequential run:

<img width="1728" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/27236d53-630c-418d-bc6f-60c8b925bd13">

3x less memory consumption, so yeah, it's something that requires investigation.

Reply:
@keradus @staabm I made a slight change in terms of iterator, and got rid of initial linting in parallel runner's main process where file chunk is prepared for worker. The effect is awesome: ~30MB memory usage, 2x less than for sequential analysis.

<img width="1727" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/75abe01a-bc51-459f-ab1d-d939f893056a">

The linting is done anyway, but on the workers' side, when they retrieve file chunk and iterate trough it. Basically, we've been doing that linting twice, which does not make sense, as linting result was not used on main process' side.

Thanks to this change, the overall analysis is even faster than before (previously 11.623, now 8.333) 😁.

Reply:
is the memory problem fixed on the latest commit of this PR?

Reply:
@staabm yes, it should be fixed in [c1ce065](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777/commits/c1ce0657535bd0a6383af0626151e89152589712) 🙂. Tested it on 30k+ repo without bumping memory limit and it works.

Reply:
Cool. It might still make sense to create a profile from a worker procesd

Reply:
nice investigation 👍🏻 

for solution that I see, if I understand right it will not only improve memory usage, but also reduce by half amount of times we do file linting? :)